### ISSUE TITLE
Error in color map optimization pipeline.

### ISSUE THREAD

**Issue by dalnoguer** [isl-org/Open3D#4776]
### Checklist

- [X] I have searched for [similar issues](https://github.com/isl-org/Open3D/issues).
- [X] For Python issues, I have tested with the [latest development wheel](http://www.open3d.org/docs/latest/getting_started.html#development-version-pip).
- [X] I have checked the [release documentation](http://www.open3d.org/docs/release/) and the [latest documentation](http://www.open3d.org/docs/latest/) (for `master` branch).

### My Question

Hello,

Thanks for providing such an amazing library :)

I was trying to run the color map optimization pipeline but encountered this error message.

```
[KDTreeFlann::SetRawData] Failed due to no data.
```

I have previously generated a sequence of rgbd frames and created a mesh using rgbd integration.
I have double check camera and intrinsics and extrinsics look good. I assume extrinsics are in the form world_T_camera. Is this a correct assumption?

Otherwise, I don't know what is wrong.

Thanks for your time!

Here is the piece of code I have used:

```
import pathlib

import open3d as o3d
from tqdm import tqdm

if __name__ == "__main__":

    dataset_dir = pathlib.Path("/home/dnoguer/rgbd")
    image_paths = sorted([p for p in dataset_dir.glob("image/*.png")])
    depth_paths = sorted([p for p in dataset_dir.glob("depth/*.png")])

    camera = o3d.io.read_pinhole_camera_trajectory(
        str(dataset_dir.joinpath("camera_trajectory.json"))
    )

    # Read RGBD frames
    rgbd_images = []
    for i, (image_path, depth_path) in tqdm(enumerate(zip(image_paths, depth_paths))):
        color = o3d.io.read_image(str(image_path))
        depth = o3d.io.read_image(str(depth_path))
        rgbd = o3d.geometry.RGBDImage.create_from_color_and_depth(
            color, depth, depth_trunc=1.5, convert_rgb_to_intensity=False
        )
        rgbd_images.append(rgbd)

    # Read mesh
    mesh = o3d.io.read_triangle_mesh(str(dataset_dir.joinpath("mesh.ply")))

    mesh, camera = o3d.pipelines.color_map.run_non_rigid_optimizer(
        mesh,
        rgbd_images,
        camera,
        o3d.pipelines.color_map.NonRigidOptimizerOption(
            maximum_iteration=300, maximum_allowable_depth=1.5
        ),
    )

    o3d.io.write_triangle_mesh(str(dataset_dir.joinpath("mesh_text.ply")), mesh)
```

**Comment by theNded** [comment]
World to camera is correct, but there could be other issues. For instance, is 1.5 a good clipping distance? More log info and/or some example data might be helpful.

**Comment by dalnoguer** [comment]
There are no more logs unfortunately.

Is there a way to share the data with you?

**Comment by theNded** [comment]
Could you please add `o3d.utility.set_verbosity_level(o3d.utility.VerbosityLevel.Debug)` at the beginning to provide possibly more logs?
You may share it on Google drive or email me if you like to keep it private. I cannot guarantee a very prompt solution since there are several dues though.

**Comment by dalnoguer** [comment]
```
[Open3D DEBUG] [cam 248]: 0/917868 (0.00000%) vertices are visible
[Open3D DEBUG] [cam 249]: 0/917868 (0.00000%) vertices are visible
[Open3D DEBUG] [cam 250]: 0/917868 (0.00000%) vertices are visible
[Open3D DEBUG] [cam 251]: 0/917868 (0.00000%) vertices are visible
[Open3D DEBUG] [ColorMapOptimization] Non-Rigid Optimization
[Open3D DEBUG] [Iteration 0001]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [Iteration 0002]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [Iteration 0003]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [Iteration 0004]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [Iteration 0005]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [Iteration 0006]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [Iteration 0007]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [Iteration 0008]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [Iteration 0009]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [Iteration 0010]
[Open3D DEBUG] Residual error : 0.000000, reg : 0.000000
[Open3D DEBUG] [ColorMapOptimization] Set Mesh Color
[Open3D DEBUG] [RemoveDuplicatedVertices] 0 vertices have been removed.
[Open3D DEBUG] [RemoveDuplicatedTriangles] 0 triangles have been removed.
[Open3D DEBUG] [RemoveUnreferencedVertices] 0 vertices have been removed.
[Open3D DEBUG] [RemoveDegenerateTriangles] 0 triangles have been removed.
[Open3D DEBUG] Triangle mesh sampled from 917868 vertices and 931659 triangles to 0 vertices and 0 triangles.
[Open3D WARNING] [KDTreeFlann::SetRawData] Failed due to no data.
[Open3D DEBUG] Write geometry::TriangleMesh: 931659 triangles and 917868 vertices.
```

It seems extrinsics are off. Will double check on my side, thanks.
Is there a way to obtain the masks mentioned here?

```[Open3D DEBUG] [ColorMapOptimization] CreateDepthBoundaryMasks
[Open3D DEBUG] [MakeDepthMasks] geometry::Image 0/252
[Open3D DEBUG] [MakeDepthMasks] geometry::Image 1/252
[Open3D DEBUG] [MakeDepthMasks] geometry::Image 2/252
[Open3D DEBUG] [MakeDepthMasks] geometry::Image 3/252
[Open3D DEBUG] [MakeDepthMasks] geometry::Image 4/252
[Open3D DEBUG] [MakeDepthMasks] geometry::Image 5/252
[Open3D DEBUG] [MakeDepthMasks] geometry::Image 6/252
[Open3D DEBUG] [MakeDepthMasks] geometry::Image 7/252
```

Thanks!

**Comment by dalnoguer** [comment]
![image](https://user-images.githubusercontent.com/26818829/154699779-557231e0-186e-4171-bd70-e3b91cf1392a.png)

Camera centers are correct.

**Comment by theNded** [comment]
Is `depth_scale` 1000 for your sensor? By default we use that to covnert depth from uint16 to float, but that may not hold for different sensors

**Comment by dalnoguer** [comment]
Yes, I save depth in uint16 in mm. 
I will send the data over gdrive and email.

Thanks

**Comment by theNded** [comment]
You may reach me at [weidong at andrew dot cmu dot edu]
But before doing that, could you please do another check by
1. Load the mesh;
2. Load one (or several RGBD images) then create point clouds with [this API](http://www.open3d.org/docs/release/python_api/open3d.geometry.PointCloud.html#open3d.geometry.PointCloud.create_from_rgbd_image) passing in your extrinsic;
3. Visualize them together with `draw`.
and see if they align well? Do not change any other parameters you are using now to make sure the same behavior.

**Comment by dalnoguer** [comment]
I have loaded the mesh and it aligns well with a lifted pointcloud.

I debugged the cpp code and the issue was coming from depth - rgb alignment. My camera setup does not provide aligned frames, therefore when aligning depth with rgb, the depth was left with a lot of holes. Then, in [this condition](https://github.com/isl-org/Open3D/blob/master/cpp/open3d/pipelines/color_map/ColorMapUtils.cpp#L182-L183) failed, since the depth sensor was 0 (it hit a hole). I applied a depth filling algorithm and now it runs well.

Thanks for your help!
